scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -323,11 +344,184 @@ def highly_variable_genes(
+        If batch_key is given, this denotes in how many batches genes are detected as HVG
+    """
+
+    try:
+        from skmisc.loess import loess
+    except ImportError:
+        raise ImportError(
+            'Please install skmisc package via `pip install --user scikit-misc'
+        )
+
+    if batch_key is None:
+        batch_info = pd.Categorical(np.zeros(adata.shape[0], dtype=int))
+    else:
+        batch_info = adata.obs[batch_key]
+
+    norm_gene_vars = []
+    for b in np.unique(batch_info):
+
+        mean, var = _get_mean_var(adata[batch_info == b].X)
+        not_const = var > 0
+        estimat_var = np.zeros(adata.shape[1], dtype=np.float64)
+
+        y = np.log10(var[not_const])
+        x = np.log10(mean[not_const])
+        model = loess(x, y, span=span, degree=2)
+        model.fit()
+        estimat_var[not_const] = model.outputs.fitted_values
+        reg_std = np.sqrt(10 ** estimat_var)
+
+        batch_counts = adata[batch_info == b].X.astype(np.float64).copy()
+        # clip large values as in Seurat
+        N = np.sum(batch_info == b)
+        vmax = np.sqrt(N)
+        clip_val = reg_std * vmax + mean
+        if sp_sparse.issparse(batch_counts):
+            batch_counts = sp_sparse.csr_matrix(batch_counts)
+            mask = batch_counts.data > clip_val[batch_counts.indices]
+            batch_counts.data[mask] = clip_val[batch_counts.indices[mask]]
+        else:
+            clip_val_broad = np.broadcast_to(clip_val, batch_counts.shape)
+            np.putmask(
+                batch_counts, batch_counts > clip_val_broad, clip_val_broad,
+            )
+
+        if sp_sparse.issparse(batch_counts):
+            squared_batch_counts_sum = np.array(batch_counts.power(2).sum(axis=0))
+            batch_counts_sum = np.array(batch_counts.sum(axis=0))
+        else:
+            squared_batch_counts_sum = np.square(batch_counts).sum(axis=0)
+            batch_counts_sum = batch_counts.sum(axis=0)
+
+        norm_gene_var = (1 / ((N - 1) * np.square(reg_std))) * (
+            (N * np.square(mean))
+            + squared_batch_counts_sum
+            - 2 * batch_counts_sum * mean
+        )
+        norm_gene_vars.append(norm_gene_var.reshape(1, -1))
+
+    norm_gene_vars = np.concatenate(norm_gene_vars, axis=0)
+    # argsort twice gives ranks
+    ranked_norm_gene_vars = np.argsort(np.argsort(norm_gene_vars, axis=1), axis=1)
+    median_ranked = np.median(ranked_norm_gene_vars, axis=0)
+
+    num_batches_high_var = np.sum(
+        ranked_norm_gene_vars >= (adata.X.shape[1] - n_top_genes), axis=0
+    )
+    df = pd.DataFrame(index=np.array(adata.var_names))
+    df["highly_variable_nbatches"] = num_batches_high_var
+    df["highly_variable_rank"] = median_ranked
+    df["variances_norm"] = np.mean(norm_gene_vars, axis=0)
+    df["means"] = mean
+    df["variances"] = var
+
+    df.sort_values(
+        ["highly_variable_nbatches", "highly_variable_rank"],
+        ascending=False,
```

Comment:
Should this be `[False, True]`? Not clear if bigger highly_variable_rank value means less or more variable.

Reply:
higher rank means higher normalized variance (i.e., more variable)